limits/filetransfer/limits.go

```diff
@@ -0,0 +1,391 @@
+	}
+}
+
+func (p *Permit) PreserveModes(path string, mode os.FileMode) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	if (mode & S_ISUID) != 0 {
+		mode &= os.ModePerm
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
+		}
+		mode |= os.ModeSetuid
+	} else {
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
+		}
+		mode &= os.ModePerm
+	}
+
+	if !p.limits.FileTransfer.DoNotPreserveMode {
+		return os.Chmod(path, mode)
+	} else {
+		return nil
+	}
+}
+
+func (p *Permit) PreserveOwnerGroup(path string, uid int, gid int) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	forcedSet := false
+	if p.limits.FileTransfer.OwnerPut != "" {
+		u, err := user.Lookup(p.limits.FileTransfer.OwnerPut)
+		if err != nil {
+			return err
+		}
+		uid, err = strconv.Atoi(u.Uid)
+		if err != nil {
+			return err
+		}
+		forcedSet = true
+	}
+	if p.limits.FileTransfer.GroupPut != "" {
+		u, err := user.LookupGroup(p.limits.FileTransfer.GroupPut)
+		if err != nil {
+			return err
+		}
+		gid, err = strconv.Atoi(u.Gid)
+		if err != nil {
+			return err
+		}
+		forcedSet = true
+	}
+	if forcedSet {
+		return os.Chown(path, uid, gid)
+	}
+	if !p.limits.FileTransfer.DoNotPreserveOwner {
+		return os.Chown(path, uid, gid)
+	} else {
+		return nil
+	}
+}
+
+func updatePerHourCounters() {
+	if counterUpdateRunning {
+		counterUpdateStarted <- false
+		return
+	}
+
+	counterUpdateRunning = true
+	counterUpdateStarted <- true
+	for counterUpdateRunning {
+		for minute := 0; minute < 60; minute++ {
+			time.Sleep(countersUpdateSleepTimeS)
+			countersMutex.Lock()
+			if deviceCountersLastH.period >= math.MaxUint32-1 {
+				deviceCountersLastH.period = 0
+			}
+			deviceCountersLastH.period++
+			sinceLastUpdateS := time.Now().Unix() - deviceCountersLastH.bytesTransferredLastUpdate.Unix()
+			if deviceCountersLastH.bytesTransferred != 0 {
+				deviceCountersLastH.currentTxRate = float64(deviceCountersLastH.bytesTransferred*1.0) / float64(sinceLastUpdateS)
+			}
+			sinceLastUpdateS = time.Now().Unix() - deviceCountersLastH.bytesReceivedLastUpdate.Unix()
+			if deviceCountersLastH.bytesReceived != 0 {
+				deviceCountersLastH.currentRxRate = float64(deviceCountersLastH.bytesReceived*1.0) / float64(sinceLastUpdateS)
+			}
+			countersMutex.Unlock()
+		}
+		countersMutex.Lock()
+		deviceCountersLastH.bytesTransferredLastH = deviceCountersLastH.bytesTransferred
+		deviceCountersLastH.bytesReceivedLastH = deviceCountersLastH.bytesTransferred
+		deviceCountersLastH.bytesTransferred = 0
+		deviceCountersLastH.bytesReceived = 0
+		deviceCountersLastH.currentRxRate = 0.0
+		deviceCountersLastH.currentTxRate = 0.0
```

Comment:
I kind of thought that it would be little value, since most of the devices transfer little data, so resetting every hour seemed the least costly (both implementation wise and CPU wise) but let me see if I can fix that as you say.

Reply:
changed to exponentially weighted moving average (updated every countersUpdateSleepTimeS seconds and with average taken over 1 minute (very similar to what uptime does)).